test/syscalls/linux/mq.cc

```diff
@@ -0,0 +1,283 @@
+  }
+  return PosixQueue(fd, name);
+}
+// MqOpen wraps mq_open(3) using a generated name.
+PosixErrorOr<PosixQueue> MqOpen(std::string name, int oflag, mode_t mode,
+                                struct mq_attr* attr) {
+  mqd_t fd = mq_open(name.c_str(), oflag, mode, attr);
+  if (fd == -1) {
+    return PosixError(errno, absl::StrFormat("mq_open(%d)", oflag));
+  }
+  return PosixQueue(fd, name);
+}
+
+// MqUnlink wraps mq_unlink(2).
+PosixError MqUnlink(std::string name) {
+  int err = mq_unlink(name.c_str());
+  if (err == -1) {
+    return PosixError(errno, absl::StrFormat("mq_unlink(%s)", name.c_str()));
+  }
+  return NoError();
+}
+
+// MqClose wraps mq_close(2).
+PosixError MqClose(mqd_t fd) {
+  int err = mq_close(fd);
+  if (err == -1) {
+    return PosixError(errno, absl::StrFormat("mq_close(%d)", fd));
+  }
+  return NoError();
+}
+
+// Test simple opening and closing of a message queue.
+TEST(MqTest, Open) {
+  SKIP_IF(IsRunningWithVFS1());
+
+  PosixQueue queue = ASSERT_NO_ERRNO_AND_VALUE(
+      MqOpen(O_RDWR | O_CREAT | O_EXCL, 0777, nullptr));
+}
+
+// Test mq_open(2) after mq_unlink(2).
+TEST(MqTest, OpenAfterUnlink) {
+  GTEST_SKIP();
+
+  PosixQueue queue = ASSERT_NO_ERRNO_AND_VALUE(
+      MqOpen(O_RDWR | O_CREAT | O_EXCL, 0777, nullptr));
+
+  ASSERT_NO_ERRNO(MqUnlink(queue.name()));
+  EXPECT_THAT(MqOpen(queue.name(), O_RDWR), PosixErrorIs(ENOENT));
+  ASSERT_NO_ERRNO(MqClose(queue.release()));
+}
+
+// Test using invalid args with mq_open.
+TEST(MqTest, OpenInvalidArgs) {
+  SKIP_IF(IsRunningWithVFS1());
+
+  // Name must start with a slash.
+  EXPECT_THAT(MqOpen("test", O_RDWR), PosixErrorIs(EINVAL));
+
+  // Name can't contain more that one slash.
+  EXPECT_THAT(MqOpen("/test/name", O_RDWR), PosixErrorIs(EACCES));
+
+  // mq_attr's mq_maxmsg and mq_msgsize must be > 0.
+  struct mq_attr attr;
+  attr.mq_maxmsg = -1;
+  attr.mq_msgsize = 10;
+
+  EXPECT_THAT(MqOpen(O_RDWR | O_CREAT | O_EXCL, 0777, &attr),
+              PosixErrorIs(EINVAL));
+
+  attr.mq_maxmsg = 10;
+  attr.mq_msgsize = -1;
+
+  EXPECT_THAT(MqOpen(O_RDWR | O_CREAT | O_EXCL, 0777, &attr),
+              PosixErrorIs(EINVAL));
+
+  // Names should be shorter than NAME_MAX.
+  char max[NAME_MAX + 3];
+  max[0] = '/';
+  for (size_t i = 1; i < NAME_MAX + 2; i++) {
+    max[i] = 'a';
+  }
+  max[NAME_MAX + 2] = '\0';
+
+  EXPECT_THAT(MqOpen(std::string(max), O_RDWR | O_CREAT | O_EXCL, 0777, &attr),
+              PosixErrorIs(ENAMETOOLONG));
+}
+
+// Test creating a queue that already exists.
+TEST(MqTest, CreateAlreadyExists) {
+  SKIP_IF(IsRunningWithVFS1());
+
+  PosixQueue queue = ASSERT_NO_ERRNO_AND_VALUE(
+      MqOpen(O_RDWR | O_CREAT | O_EXCL, 0777, nullptr));
+
+  EXPECT_THAT(MqOpen(queue.name(), O_RDWR | O_CREAT | O_EXCL, 0777, nullptr),
+              PosixErrorIs(EEXIST));
+}
+
+// Test opening a queue that doesn't exists.
+TEST(MqTest, NoQueueExists) {
+  SKIP_IF(IsRunningWithVFS1());
+
+  EXPECT_THAT(MqOpen(O_RDWR), PosixErrorIs(ENOENT));
```

Comment:
This is a somewhat convoluted use of the overloaded implementation of `MqOpen`. This works because:
- If MqOpen isn't passed a name, it generates one
- Since we didn't pass O_CREAT, it won't create the queue

Just be explicit and pass a name we know doesn't exist: 
```
EXPECT_THAT(MqOpen("/gvisor-mq-test-nonexistent-queue", O_RDWR), PosixErrorIs(ENOENT));
```

Reply:
Done